### ISSUE TITLE
Can we remove TextFormat.textColor?

### ISSUE THREAD

**Issue by dandclark** [w3c/edit-context#54]
There is a problem with [`TextFormat.textColor`](https://w3c.github.io/edit-context/#ref-for-dom-textformat-textcolor-1) in that it's not clear what value this field should contain when the text color does not need to be changed. Since the `textColor` field is always present, an author can't tell the difference between a `TextFormatUpdateEvent` instructing the page to both show an underline and to change the text color, and a `TextFormatUpdateEvent` that is only instructing the page to show an underline.

One solution to fix this is to specify that an empty string means the author should not modify the text color.

Stepping back, however, it's not clear this field is needed at all. @snianu informed me of the history here, where this was added as a potential requirement for Windows dicatation, where text color is used to indicate the system's confidence in a candidate word. However, this hasn't yet been implemented and it's not clear whether there are any other definite use cases.

Therefore I think the simplest choice is to just remove the `textColor` field until we have a proven need for it. If we remove it now before EditContext ships, then it's easy to add it back -- but if we ship the field without a good understanding of how it will be used then we may find it hard to make needed changes later on.

**Comment by dandclark** [comment]
Notes from 10/8/2023 EditingWG meeting:

> 6:43 PM "Can we remove TextFormat.textColor?"
6:44 PM dandclark: IME may want to format the text
6:44 PM dandclark: textColor is one of those
6:45 PM dandclark: most of the the event doesn't inform about the expected color
6:46 PM dandclark: there was a plan to have a OS level feature, but it hasn't been implemented
6:47 PM dandclark: so, remove until there is a use case, and if added back later, make sure it is clear to devs when/how to use it
6:48 PM whsieh: one motion was also Mac's touchbar
6:49 PM johanneswilm: what about using EditContext with canvas?
6:49 PM dandclark: canvas should still get beforeinput
6:52 PM smaug: there are other properties too, like backgroundColor
6:53 PM dandclark: they do have similar issues too
6:55 PM whsieh: I think removing them and adding them back later is the way to go
6:56 PM dandclark: remove textColor, and as a followup action item I'll check whether there are use cases for other properties
6:57 PM resolution: see above ^

**Comment by masayuki-nakano** [comment]
Wow, I'm surprised at this change. For some IMEs, the foreground/background colors are important. So, this makes some IME users' UX of EditContext on Firefox (completely support the native LnF) much worse than `contenteditable` or text control element.

E.g., ATOK on Windows looks like:
<img width="206" height="76" alt="Image" src="https://github.com/user-attachments/assets/918a0732-0e14-4f34-815e-4dcf8d4dfaa9" />

If the compositions string contains different kind of string, the composition string is painted as:
<img width="824" height="240" alt="Image" src="https://github.com/user-attachments/assets/c2bc1eeb-b13c-43f9-b7ba-a9ad70fb1b2d" />

And if the composition string has some warnings, it looks like:
<img width="846" height="104" alt="Image" src="https://github.com/user-attachments/assets/645b4d6a-9b44-46cb-8f11-a4836d0a87f8" />
(Only the first 6 characters are the composition string, the latters are generated by IME to warn the wrong point within the composition string.)

**Comment by masayuki-nakano** [comment]
And perhaps, that makes it harder to see which clause is selected for some IME users because IMEs may change only the colors with same underline style.

**Comment by smaug----** [comment]
So does EditContext effectively end up already leaking some information about the IME user is using? Could we fix that somehow without making the UX worse?
And I wonder whether this overlaps with the proposed HTML-in-Canvas, if one uses contenteditable there.

**Comment by ltenmoz** [comment]
Even without colors, the information about the style and thickness of the underline is currently leaked, which could narrow down which IME is being used in some cases. For DOM-based EditContext, the API could be changed so that the composition range in the DOM is reported by the web app to the UA, which applies styling to it. But that doesn't work at all for canvas-based EditContext - I really don't see any way of not leaking some information about the IME for that (without harming UX).

I filed https://github.com/WICG/html-in-canvas/issues/143 about HTML-in-canvas. For that, we can potentially paint the IME underline "on top" of the canvas (so it can't be read back),  although that doesn't really work for text/background colors (so browsers would have to use some fallback for that if they want to not leak information). For EditContext it's more difficult, since the web app fully controls the text rendering.

**Comment by smaug----** [comment]
> For that, we can potentially paint the IME underline "on top" of the canvas

That sounds tricky if one applies some transforms and/or paints something else on top of the text

**Comment by ltenmoz** [comment]
Yes indeed. In theory the position of the element should be kept in sync with where it's being rendered anyways for accessibility, but  it still could be an issue

**Comment by ltenmoz** [comment]
Okay looking through the telemetry data we have, I would estimate that around 5% of Desktop Windows IME users would be affected by this (of course, this will likely vary between browsers due to different geographic distribution of users). In addition to ATOK mentioned by Masayuki, other notable IMEs which use color are
- Microsoft Bopomofo (zh-TW) which highlights part of the composition with a background color to indicate an invalid composition. The entire composition is underlined, so replacing the color with an underline loses information.
- iFlyTek Input Method (zh-CN) which highlights the entire composition with a foreground color (so, no information would be lost by using an underline instead, but it may be unexpected to the user)
- Palm Input Method (zh-CN) which does the same as iFlyTek, if the experimental "single-line mode" setting is enabled
- Hindi Indic Input 3 (hi-IN) also highlights entire composition with foreground color

I'm not sure to what degree this represents a new fingerprinting vector, since you can probably narrow down the IME quite a bit as it is by analyzing key/beforeinput/composition events (for example, ATOK is probably unique in having the exact warning message shown above), although it does require more effort. Fingerprinting concerns aside, personally I think it would be best to support this, since 5% of IME users is still a large number of people.

**Comment by johanneswilm** [comment]
Call 2026-07-09:

> Leo Tenenbaum proposed adding text color and background color properties 
to text format update events to better support IME highlighters.
Dan Clark noted that this feature was previously removed due to lack of 
known use cases but agreed that adding it back is reasonable if there is 
a demonstrated need.
Leo Tenenbaum will investigate usage statistics and potential privacy 
concerns related to fingerprinting before finalizing this change.
Xiaoqian Wu inquired about inviting participants from the HTML-in-canvas 
project to the group. Johannes Wilm and Olli Pettay clarified that while 
the group's current charter excludes editing in canvas—as it involves 
custom drawing rather than standard primitives—they remain open to 
collaboration if canvas-based editors face issues that intersect with 
the group's work, particularly regarding privacy and IME implementation.

**Comment by smaug----** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=2049462#c3

**Comment by dandclark** [comment]
Notes from the last call:

> Olli reported feedback from Mozilla's security and privacy reviewers.
Their current position is that accessibility and user experience should
take precedence over the potential fingerprinting concerns associated
with exposing the actual IME text colors.
The proposed direction would therefore be to restore the ability to
expose the actual text colors in the EditContext specification. If users
require stronger privacy protection, the implementation could instead
return default or generic colors.
Dan agreed that this direction appears reasonable based on the available
research, but noted that it would still require a privacy review on the
Chromium side. He did not expect Chromium's privacy review to reach a
substantially different conclusion.
The group also discussed the scope of the affected IMEs. The current
feedback appears to concern specific Windows IMEs, although there was
some uncertainty about whether similar behavior exists on Linux or
macOS. Johannes noted that Linux IMEs can also have configurable
highlight colors, suggesting that the issue may not necessarily be
Windows-specific.
Wenson brought up an analogy with how Safari/WebKit handles system
accent colors: the actual customized system color is used for rendering,
while script-visible values can be normalized to a default value.
This led to a potential alternative design for EditContext: instead of
exposing the actual IME color to the page, the specification could
define an opaque or special CSS color value representing the IME
highlight color. The page would be able to use this value for rendering
without being able to determine the underlying system-specific color.
> 
> 
> Dan and Olli found this idea promising and agreed that it was worth
investigating further. The approach could potentially provide the
necessary rendering capability while limiting the fingerprinting
surface.
Johannes raised a concern that the group may not have enough expertise
with the wide variety of IMEs across different operating systems. He
noted that IME highlight colors can also be configured on Linux, so the
group should avoid making assumptions based solely on Windows behavior.
Olli suggested relying on Masayuki for additional input, given his
extensive experience testing different keyboards and IMEs across
multiple operating systems.

**Comment by ltenmoz** [comment]
There is the issue with using an opaque CSS color value that for <canvas> EditContext, we would have to disable pixel readback and it wouldn't work with WebGL/WebGPU.
